src/Gate.jl

```diff
@@ -162,160 +272,145 @@ const U1 = Rz
 
 A controlled gate.
 """
-abstract type Control{Op<:Operator} <: Operator{NamedTuple{(:target,),Tuple{Operator}}} end
+struct Control{Op<:Operator} <: Operator
+    target::Op
+end
+
+Control(op::Op) where {Op} = Control{Op}(op)
+Control{Op}(; params...) where {Op} = Control{Op}(Op(; params...))
+
+Control(lane, target::Gate{Op,N}) where {Op,N} =
+    Gate{Control{Op},N + 1}(tuple(lane, lanes(target)...), Control(operator(target)))
+Control{Op}(lane, lanes::Vararg{Int,N}; params...) where {Op,N} =
+    Gate{Control{Op},N + 1}(tuple(lane, lanes...), Control{Op}(; params...))
+
 Base.length(::Type{Control{T}}) where {T<:Operator} = 1 + length(T)
-parameters(::Type{Control{Op}}) where {Op} = parameters(Op)
 
-"""
-    SU(N)(lane_1, lane_2, ..., lane_log2(N))
+isparametric(::Type{<:Control{T}}) where {T<:Operator} = isparametric(T)
+isparametric(op::Control{T}) where {T<:Operator} = isparametric(T)
+parameters(::Type{Control{Op}}) where {Op} = parameters(Op)
+parameters(op::Control{Op}) where {Op} = parameters(op.target)
 
-The `SU{N}` multi-qubit general unitary gate that can be used to represent any unitary matrix that acts on
-`log2(N)` qubits. A new random `SU{N}` can be created with `rand(SU{N}, lanes...)`, where `N` is the dimension
- of the unitary matrix and `lanes` are the qubit lanes on which the gate acts.
-"""
-abstract type SU{N} <: Operator{NamedTuple{(:array,), Tuple{Matrix}}} end
-Base.length(::Type{SU{N}}) where {N} =
-    ispow2(N) ? log2(N) |> Int : throw(DomainError(N, "N must be a power of 2"))
+Base.adjoint(::Type{Control{Op}}) where {Op} = Control{adjoint(Op)}
+Base.adjoint(op::Control{Op}) where {Op} =
+    if isparametric(op)
+        Control{Op'}(operator(op)')
+    else
+        Control{Op'}()
+    end
 
 for Op in [:X, :Y, :Z, :Rx, :Ry, :Rz]
     @eval const $(Symbol("C" * String(Op))) = Control{$Op}
 end
 
-# adjoints
-for Op in [:I, :X, :Y, :Z, :Rx, :Ry, :Rz, :Rxx, :Ryy, :Rzz, :U2, :U3, :H, :Swap, :FSim]
-    @eval Base.adjoint(::Type{$Op}) = $Op
-end
+# Gate{Control}
+targettype(::Type{Op}) where {Op<:Operator} = Op
+targettype(::Type{Control{Op}}) where {Op} = Op
+targettype(::Type{Control{Op}}) where {Op<:Control} = targettype(Op)
+targettype(::Type{<:Gate{Op}}) where {Op} = targettype(Op)
 
-Base.adjoint(::Type{S}) = Sd
-Base.adjoint(::Type{Sd}) = S
-Base.adjoint(::Type{T}) = Td
-Base.adjoint(::Type{Td}) = T
+control(g::G) where {G<:Gate{<:Control}} = lanes(g)[1:end-length(targettype(G))]
+target(g::G) where {G<:Gate{<:Control}} = lanes(g)[end-length(targettype(G))+1:end]
 
-Base.adjoint(::Type{Control{Op}}) where {Op} = Control{adjoint(Op)}
+"""
+    SU{N}(lane_1, lane_2, ..., lane_N, array)
 
-# operator sets
-const Pauli = Union{I,X,Y,Z}
-const Phase = Union{I,Z,S,Sd,T,Td,Rz,Hz,FSim}
+The `SU{N}` multi-qubit general unitary gate that can be used to represent any unitary matrix that acts on
+`N` qubits. A new random `SU{N}` can be created with `rand(SU{N}, lanes...)`, where `N` is the dimension
```

Comment:
```suggestion
`log2(N)` qubits. A new random `SU{N}` can be created with `rand(SU{N}, lanes...)`, where `N` is the dimension
```
The definition of `SU{N}` (random unitary) is that `N` is the number of dimensions of the operator, therefore, in the case of qubits, this would be `log2(N)`. This is how I have seen `SU` defined everywhere (see in [`Qiskit`](https://docs.quantum.ibm.com/api/qiskit/0.19/qiskit.quantum_info.random_unitary), in some [papers](https://arxiv.org/pdf/1811.12926.pdf), ...).

Reply:
Yep, but using `N` as the dimension of the matrix is super annoying. First, it doesn't tell you in how many sites it acts. Also I added a note to describe that `N` in our case is the number of sites, not the dimensionality.